transport/transport-runtime/src/main/java/com/google/android/datatransport/runtime/scheduling/jobscheduling/Uploader.java

```diff
@@ -14,13 +14,98 @@
 
 package com.google.android.datatransport.runtime.scheduling.jobscheduling;
 
+import android.content.Context;
+import android.net.ConnectivityManager;
+import android.net.NetworkInfo;
+import com.google.android.datatransport.runtime.BackendRegistry;
+import com.google.android.datatransport.runtime.BackendRequest;
+import com.google.android.datatransport.runtime.BackendResponse;
+import com.google.android.datatransport.runtime.EventInternal;
+import com.google.android.datatransport.runtime.TransportBackend;
+import com.google.android.datatransport.runtime.scheduling.persistence.EventStore;
+import com.google.android.datatransport.runtime.scheduling.persistence.PersistedEvent;
+import com.google.android.datatransport.runtime.synchronization.SynchronizationGuard;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Executor;
 import javax.inject.Inject;
 
 /** Handles upload of all the events corresponding to a backend. */
 public class Uploader {
 
+  private final Context context;
+  private final BackendRegistry backendRegistry;
+  private final EventStore eventStore;
+  private final WorkScheduler workScheduler;
+  private final Executor executor;
+  private final SynchronizationGuard guard;
+
   @Inject
-  public Uploader() {}
+  public Uploader(
+      Context context,
+      BackendRegistry backendRegistry,
+      EventStore eventStore,
+      WorkScheduler workScheduler,
+      Executor executor,
+      SynchronizationGuard guard) {
+    this.context = context;
+    this.backendRegistry = backendRegistry;
+    this.eventStore = eventStore;
+    this.workScheduler = workScheduler;
+    this.executor = executor;
+    this.guard = guard;
+  }
+
+  boolean isNetworkAvailable() {
+    ConnectivityManager connectivityManager =
+        (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
+    NetworkInfo activeNetworkInfo = connectivityManager.getActiveNetworkInfo();
+    return activeNetworkInfo != null && activeNetworkInfo.isConnected();
+  }
+
+  void upload(String backendName, int attemptNumber, Runnable callback) {
+    this.executor.execute(
+        () -> {
+          try {
+            if (!isNetworkAvailable()) {
+              workScheduler.schedule(backendName, attemptNumber + 1);
+            } else {
+              logAndUpdateState(backendName, attemptNumber);
+            }
+          } finally {
+            callback.run();
+          }
+        });
+  }
+
+  void logAndUpdateState(String backendName, int attemptNumber) {
+    TransportBackend backend = backendRegistry.get(backendName);
+    List<EventInternal> eventInternals = new ArrayList<>();
+    Iterable<PersistedEvent> persistedEvents =
```

Comment:
If I understand correctly, the explicit synchronization is only required because we want to support multi process thread safety ?

Questions
1. The API for the datastore seems to require that call sites responsibly call eventStore methods in critical sections. I see that  this cannot be done within the implementations of the eventStore since we invoke others apis from within critical sections, that are outside of the store. Would a single operation on datastore have to be run in a critical section too?
2. If the critical sections are scattered across the project, I would vote to bring them all into a single file and encapsulate them as higher level functions

Reply:
> require that call sites responsibly call eventStore methods in critical sections

It's not the case, you can call them by themselves they will just be part of their own isolated transaction, the reason you'd want to call them within a critical section is to guarantee atomicity.

2. iiu your comment correctly `runCriticalSection` is exactly the higher-level function you're describing, so I am not sure what the suggestion is here, can you be more specific, e.g. provide an example?

Reply:
Ok. So it seems like each operation on datastore itself need not be in a critical section. THis is requires only when multiple operations are to be atomic.

2. What I mean is -> All invocations of `runCriticalSection` could be centralized. We will have a single place to reason about all higher level operations that we need to be atomic. Is this file that place or are there other files that invoke `runCtiticalSection` ?

Reply:
@VinayGuthal this is what I was referring to in 1. Namely, this is what's called as a result of `transport.send(event)`.

Reply:
Ok. Defer to your judgement